File: src/Plane.cpp
/** @file   Plane.cpp
    @author Philip Abbet
    
    Implementation of the class Athena::Math::Plane
    
    @note   This is based on the corresponding class from
            <a href="http://www.ogre3d.org/">Ogre3D</a>.
*/

#include <Athena-Math/Plane.h>
//#include <Athena-Math/Matrix3.h>
#include <Athena-Math/AxisAlignedBox.h>

using namespace Athena::Math;


/****************************** CONSTRUCTION / DESTRUCTION ******************************/

Plane::Plane()
: normal(Vector3::ZERO), d(0.0)
{
}

//-----------------------------------------------------------------------

Plane::Plane(const Plane& rhs)
: normal(rhs.normal), d(rhs.d)
{
}

//-----------------------------------------------------------------------

Plane::Plane(const Vector3& rkNormal, Real fConstant)
: normal(rkNormal), d(-fConstant)
{
}

//-----------------------------------------------------------------------

Plane::Plane(Real a, Real b, Real c, Real _d)
: normal(a, b, c), d(_d)
{
}

//-----------------------------------------------------------------------

Plane::Plane(const Vector3& rkNormal, const Vector3& rkPoint)
{
	redefine(rkNormal, rkPoint);
}

//-----------------------------------------------------------------------

Plane::Plane(const Vector3& rkPoint0, const Vector3& rkPoint1, const Vector3& rkPoint2)
{
	redefine(rkPoint0, rkPoint1, rkPoint2);
}


/**************************************** METHODS ***************************************/

Real Plane::getDistance(const Vector3& rkPoint) const
{
	return normal.dotProduct(rkPoint) + d;
}

//-----------------------------------------------------------------------

Plane::tSide Plane::getSide(const Vector3& rkPoint) const
{
	Real fDistance = getDistance(rkPoint);

	if (fDistance < 0.0)
		return Plane::NEGATIVE_SIDE;

	if (fDistance > 0.0)
		return Plane::POSITIVE_SIDE;

	return Plane::NO_SIDE;
}

//-----------------------------------------------------------------------

Plane::tSide Plane::getSide(const AxisAlignedBox& box) const
{
	if (box.isNull()) 
		return NO_SIDE;

	if (box.isInfinite())
		return BOTH_SIDE;

    return getSide(box.getCenter(), box.getHalfSize());
}

//-----------------------------------------------------------------------

Plane::tSide Plane::getSide(const Vector3& centre, const Vector3& halfSize) const
{
    Real dist = getDistance(centre);

    Real maxAbsDist = normal.absDotProduct(halfSize);

    if (dist < -maxAbsDist)
        return Plane::NEGATIVE_SIDE;

    if (dist > +maxAbsDist)
        return Plane::POSITIVE_SIDE;

    return Plane::BOTH_SIDE;
}

//-----------------------------------------------------------------------

void Plane::redefine(const Vector3& rkPoint0, const Vector3& rkPoint1,
                     const Vector3& rkPoint2)
{
	Vector3 kEdge1 = rkPoint1 - rkPoint0;
	Vector3 kEdge2 = rkPoint2 - rkPoint0;
	normal = kEdge1.crossProduct(kEdge2);
	normal.normalise();
	d = -normal.dotProduct(rkPoint0);
}

//-----------------------------------------------------------------------

void Plane::redefine(const Vector3& rkNormal, const Vector3& rkPoint)
{
	normal = rkNormal;
	d = -rkNormal.dotProduct(rkPoint);
}

//-----------------------------------------------------------------------
/*
Vector3 Plane::projectVector(const Vector3& p) const
{
    normalize();
    
	Matrix3 xform;
	xform[0][0] = 1.0f - normal.x * normal.x;
	xform[0][1] = -normal.x * normal.y;
	xform[0][2] = -normal.x * normal.z;
	xform[1][0] = -normal.y * normal.x;
	xform[1][1] = 1.0f - normal.y * normal.y;
	xform[1][2] = -normal.y * normal.z;
	xform[2][0] = -normal.z * normal.x;
	xform[2][1] = -normal.z * normal.y;
	xform[2][2] = 1.0f - normal.z * normal.z;

	return xform * p;
}
*/
//-----------------------------------------------------------------------

Real Plane::normalise(void)
{
    Real fLength = normal.length();

    // Will also work for zero-sized vectors, but will change nothing
    if (fLength > 1e-08f)
    {
        Real fInvLength = 1.0f / fLength;
        normal *= fInvLength;
        d *= fInvLength;
    }

    return fLength;
}

File: scripting/Plane.js
/** @file   Plane.js
    @author Philip Abbet

    Declaration of the class Athena.Math.Plane

    @note   Using a pure JavaScript class here instead of a binding over
            the C++ one results in faster code.
*/

Athena.Math.Plane = function()
{
    if (arguments.length == 0)
    {
        this.normal = new Athena.Math.Vector3();
        this.d = 0.0;
    }
    else if ((arguments.length == 1) && (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Plane'))
    {
        this.normal = arguments[0].normal;
        this.d = arguments[0].d;
    }
    else if ((arguments.length == 2) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[1]) === 'number'))
    {
        this.normal = arguments[0];
        this.d = -arguments[1];
    }
    else if ((arguments.length == 2) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[1]) === 'object') && (arguments[1].__classname__ == 'Athena.Math.Vector3'))
    {
        this.normal = arguments[0];
        this.d = -arguments[0].dot(arguments[1]);
    }
    else if ((arguments.length == 3) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[1]) === 'object') && (arguments[1].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[2]) === 'object') && (arguments[2].__classname__ == 'Athena.Math.Vector3'))
    {
        var edge1 = arguments[1].sub(arguments[0]);
        var edge2 = arguments[2].sub(arguments[0]);
        this.normal = edge1.cross(edge2);
        this.normal.normalise();
        this.d = -this.normal.dot(arguments[0]);
    }
    else if (arguments.length == 4)
    {
        this.normal = new Athena.Math.Vector3(arguments[0], arguments[1], arguments[2]);
        this.d = arguments[3];
    }
    else
    {
        throw 'Invalid parameters, valid syntaxes:\nPlane()\nPlane(plane)\nPlane(normal, constant)\nPlane(a, b, c, d)\nPlane(normal, point)\nPlane(point0, point1, point2)';
    }
}


/*********************************** VALUE ASSIGNATION **********************************/

Athena.Math.Plane.prototype.set = function()
{
    if ((arguments.length == 1) && (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Plane'))
    {
        this.normal = arguments[0].normal;
        this.d = arguments[0].d;
    }
    else if ((arguments.length == 2) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[1]) === 'number'))
    {
        this.normal = arguments[0];
        this.d = -arguments[1];
    }
    else if ((arguments.length == 2) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[1]) === 'object') && (arguments[1].__classname__ == 'Athena.Math.Vector3'))
    {
        this.normal = arguments[0];
        this.d = -arguments[0].dot(arguments[1]);
    }
    else if ((arguments.length == 3) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[1]) === 'object') && (arguments[1].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[2]) === 'object') && (arguments[2].__classname__ == 'Athena.Math.Vector3'))
    {
        var edge1 = arguments[1].sub(arguments[0]);
        var edge2 = arguments[2].sub(arguments[0]);
        this.normal = edge1.cross(edge2);
        this.normal.normalise();
        this.d = -this.normal.dot(arguments[0]);
    }
    else if (arguments.length == 4)
    {
        this.normal = new Athena.Math.Vector3(arguments[0], arguments[1], arguments[2]);
        this.d = arguments[3];
    }
    else
    {
        throw 'Invalid parameters, valid syntaxes:\nset(plane)\nset(normal, constant)\nset(a, b, c, d)\nset(normal, point)\nset(point0, point1, point2)';
    }
}


/********************************* COMPARISON OPERATIONS ********************************/

Athena.Math.Plane.prototype.equals = function(plane)
{
    return (this.d == plane.d) && this.normal.equals(plane.normal);
}

//-----------------------------------------------------------------------

Athena.Math.Plane.prototype.notEquals = function(plane)
{
    return (this.d != plane.d) || this.normal.notEquals(plane.normal);
}


/**************************************** METHODS ***************************************/

Athena.Math.Plane.prototype.getSide = function()
{
    if ((arguments.length == 1) && (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3'))
    {
        var distance = this.getDistance(arguments[0]);

        if (distance < 0.0)
            return Athena.Math.Plane_NEGATIVE_SIDE;

        if (distance > 0.0)
            return Athena.Math.Plane_POSITIVE_SIDE;

        return Athena.Math.Plane_NO_SIDE;
    }
    else if ((arguments.length == 1) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.AxisAlignedBox'))
    {
        if (arguments[0].isNull())
            return Athena.Math.Plane_NO_SIDE;

        if (arguments[0].isInfinite())
            return Athena.Math.Plane_BOTH_SIDE;

        return this.getSide(arguments[0].getCenter(), arguments[0].getHalfSize());
    }
    else if ((arguments.length == 2) &&
             (typeof(arguments[0]) === 'object') && (arguments[0].__classname__ == 'Athena.Math.Vector3') &&
             (typeof(arguments[1]) === 'object') && (arguments[1].__classname__ == 'Athena.Math.Vector3'))
    {
        var dist = this.getDistance(arguments[0]);

        var maxAbsDist = this.normal.absDot(arguments[1]);

        if (dist < -maxAbsDist)
            return Athena.Math.Plane_NEGATIVE_SIDE;

        if (dist > +maxAbsDist)
            return Athena.Math.Plane_POSITIVE_SIDE;

        return Athena.Math.Plane_BOTH_SIDE;
    }
    else
    {
        throw 'Invalid parameters, valid syntaxes:\ngetSide(point)\ngetSide(centre, half_size)\ngetSide(axis_aligned_box)';
    }
}

//-----------------------------------------------------------------------

Athena.Math.Plane.prototype.getDistance = function(point)
{
    return this.normal.dot(point) + this.d;
}

//-----------------------------------------------------------------------

Athena.Math.Plane.prototype.projectVector = function(point)
{
    var xform = new Athena.Math.Matrix3();

    xform.m_0_0 = 1.0 - this.normal.x * this.normal.x;
    xform.m_0_1 = -this.normal.x * this.normal.y;
    xform.m_0_2 = -this.normal.x * this.normal.z;
    xform.m_1_0 = -this.normal.y * this.normal.x;
    xform.m_1_1 = 1.0 - this.normal.y * this.normal.y;
    xform.m_1_2 = -this.normal.y * this.normal.z;
    xform.m_2_0 = -this.normal.z * this.normal.x;
    xform.m_2_1 = -this.normal.z * this.normal.y;
    xform.m_2_2 = 1.0 - this.normal.z * this.normal.z;

    return xform.mul(point);
}

//-----------------------------------------------------------------------

Athena.Math.Plane.prototype.normalise = function()
{
    var length = this.normal.length();

    // Will also work for zero-sized vectors, but will change nothing
    if (length > 1e-08)
    {
        var invLength = 1.0 / length;
        this.normal.imul(invLength);
        this.d *= invLength;
    }

    return length;
}

//-----------------------------------------------------------------------

Athena.Math.Plane.prototype.toString = function()
{
    return '{ normal: ' + this.normal.toString() + ', d: ' + this.d + ' }';
}


/**************************************** OTHERS ****************************************/

Athena.Math.Plane.prototype.__classname__ = 'Athena.Math.Plane';

Athena.Math.Plane_NO_SIDE       = 0;
Athena.Math.Plane_POSITIVE_SIDE = 1;
Athena.Math.Plane_NEGATIVE_SIDE = 2;
Athena.Math.Plane_BOTH_SIDE     = 3;
